### DISCUSSION TITLE
Permissions on Docker Created Files

### DISCUSSION THREAD

**Discussion by joshso** [jmbannon/ytdl-sub#645] [Q&A] [answered]
Hi! Newbie here. (Really neat app btw Jesse!)

I run a NAS 24x7 (Unraid) and would prefer to host ytdl-sub there via Docker so that I can automate it over time. My problem is when I create files via the container, the default owner and permissions are such that I receive errors when I try to move/delete/rename using smb, connecting as my preferred username of josh.

I understand this is down to PUID/PGID/UMASK. The challenge is they look correct. Here's my work, do you spy what I'm missing?

Docker creating files as: root
I prefer Docker to be creating files as: josh

Example files and folders created from either Samba or the container:
`ls -al`

> -rw-r--r-- 1 root   root    0 Jul  4 13:44 aDockerCreatedFile
> drwxr-xr-x 1 root   root    0 Jul  4 13:49 aDockerCreatedFolder/
> -rw-rw-rw- 1 josh   users   0 Jul  4 13:35 aSambaCreatedFile
> drwxrwxrwx 1 josh   users   0 Jul  4 13:49 aSambaCreatedFolder/

This is after setting PUID=1000, PGID=100 and UMASK=000. (I cannot tell UMASK doing anything and am guessing it is setting the user side, not root?)

At least when I query umask from within the container I get the default 0022:
`umask`
> 0022

I can see the proper variable values are being passed as desired:

`docker exec ytdl-sub /usr/bin/env`
> PATH=/lsiopy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> HOSTNAME=2b735f346fab
> DOCKER_MODS=linuxserver/mods:universal-cron
> TZ=America/New_York
> HOST_OS=Unraid
> HOST_HOSTNAME=Tower
> HOST_CONTAINERNAME=ytdl-sub
> PUID=1000
> PGID=100
> UMASK=000
> PS1=$(whoami)@$(hostname):$(pwd)\$ 
> HOME=/config
> TERM=xterm
> S6_CMD_WAIT_FOR_SERVICES_MAXTIME=0
> S6_VERBOSITY=1
> S6_STAGE2_HOOK=/docker-mods
> VIRTUAL_ENV=/lsiopy
> EDITOR=nano

Here's the output of id:
`id josh`
> uid=1000(josh) gid=100(users) groups=100(users)`

While I have other docker applications running, I have never attempted to create files on the NAS from a container itself. From research I am only 80% confident that I would expect the docker created files to be owned by username=josh, group=users and with a mask of 666 (since I am setting a umask of 000). Am I wrong on expected behavior?

**Comment by jmbannon** [comment, marked as answer]
The docker container uses the username `abc`, which should be 1000:100. When you log into the container, you need to do `docker exec -u abc ytdl-sub /bin/bash`, otherwise you log in as root and it will create files using the root user

**Reply by hmc786** [reply]
Hello, 

Can you elaborate a bit more? I have a similar setup with uid=1000 gid=1000 groups=1000 and defined

     - PUID=1000
     - PGID=1000

in the docker compose. I get the same error as OP. Other containers work fine with - PUID=1000, - PGID=1000. What do you mean by "When you log into the container"?  When should I evaluate `docker exec -u abc ytdl-sub /bin/bash`?

**Reply by jmbannon** [reply]
I mean if you're going to use the CLI from within the container, you need the `-u abc` to make sure you login as the user `abc`, who gets the PUID/PGID mapping. Otherwise, you log in as root.

**Comment by nick-roux** [comment]
I'm running the ubuntu docker image on a Synology NAS.

Before I ran build docker_ubuntu to create the image, I edited Dockerfile.ubuntu by adding the following

# Below ARG DEBIAN_FRONTEND=noninteractive
# Change this to match the existing user details on your NAS 
ARG USERNAME=ytdl
ARG USER_UID=1099
ARG USER_GID=100

# Below the existing RUN command in the dockerfile
RUN useradd --uid $USER_UID --gid $USER_GID -M $USERNAME

# Below HOME="/config"
USER ytdl

# Below VOLUME /config
VOLUME /video
VOLUME /music
VOLUME /music_videos

Now your container will run under container user ytdl and access the host filesystems with host user ytdl. Anything new created in /video /music etc will be owned by user ytdl

You can also do this if you use the alpine image,, but note that the adduser command has different syntax on alpine and will need to be changed accordingly

**Reply by jmbannon** [reply]
The `abc` user will use the PUID and PGID values defined in the docker env, why not use that vs creating a new ytdl user?
